File: src/application/useCases/loadBudgetScreen.ts
import {
  averageCompletedSpend,
  budgetStatus,
  buildCategoryLines,
  detectAnomaly,
  type AverageResult,
  type BudgetStatus,
  type CategoryBudgetLine,
  type CompletedPeriodSpend,
} from '../../domain/budget'
import { categoryDistribution, assessCoverage } from '../../domain/analytics'
import { computePeriodTotals } from '../../domain/ledger'
import { dailyAllowance, type DailyAllowance } from '../../domain/analytics'
import { buildPeriod, type Period } from '../../domain/period'
import type { Halalas } from '../../domain/money'
import { withEstimatedKinds } from '../../domain/estimatedKinds'
import type { Budget, Category, CategoryBudget, Id } from '../../domain/entities/types'
import type {
  AllocationRepository,
  BudgetRepository,
  CategoryRepository,
  TransactionRepository,
} from '../ports/repositories'

/**
 * LoadBudgetScreen — spec/01.
 *
 * «سقف إجمالي وسقوف اختيارية للتصنيفات، و**متوسط منفصل من فترات مكتملة**،
 *  وعلامة شذوذ، وحالة دون تاريخ أو سقف.
 *  **لا تُخلق ميزانيات من متوسطات دون اختيار المستخدم.**»
 *
 * «منفصل» تُنفَّذ حرفيًا: `averageMinor` و `totalStatus` حقلان مختلفان،
 * ولا يُستعمل الأول قيمةً افتراضية للثاني في أي مسار.
 */

/** كم فترة سابقة تُقرأ لحساب المتوسط والشذوذ. */
const HISTORY_PERIODS = 6

export interface BudgetScreenData {
  period: Period
  budget: Budget | null
  /** null = المستخدم لم يضع سقفًا إجماليًا. */
  totalStatus: BudgetStatus | null
  spentMinor: Halalas
  /** موثوقية مصروف الفترة الجارية. */
  spentReliable: boolean
  spentNote: string | null
  /**
   * هل المصروف **معروف أصلًا**؟
   *
   * `false` حين لا تكون ولا عملية محددة النوع: وقتها `spentMinor` صفر
   * حسابيًا لكنه **مجهول** لا صفر. عرض شريط «0 من 3000» وقتها راحة
   * كاذبة، وعرض «المصروف 0.00» كذب صريح (CLAUDE.md #10).
   */
  spentKnown: boolean

  /** **معلومة تُعرض، لا سقف يُطبَّق.** */
  average: AverageResult
  anomaly: ReturnType<typeof detectAnomaly>
  allowance: DailyAllowance

  categories: Category[]
  lines: CategoryBudgetLine[]
  categoryBudgets: CategoryBudget[]
}

export interface LoadBudgetScreenDeps {
  txns: TransactionRepository
  categories: CategoryRepository
  allocations: AllocationRepository
  budgets: BudgetRepository
}

function shiftPeriod(period: Period, delta: number, payday: number): Period {
  const [year, month] = period.key.split('-').map(Number)
  const total = year * 12 + (month - 1) + delta
  return buildPeriod(Math.floor(total / 12), (total % 12) + 1, payday)
}

export function makeLoadBudgetScreen(deps: LoadBudgetScreenDeps) {
  return async function load(options: {
    period: Period
    today: string
    payday: number
  }): Promise<BudgetScreenData> {
    const { period, today, payday } = options

    const [transactions, categories, budget] = await Promise.all([
      deps.txns.listByDateRange(period.start, period.end),
      deps.categories.listAll(),
      deps.budgets.findByPeriod(period.key),
    ])
    const allocations = await deps.allocations.listByTransactionIds(transactions.map((t) => t.id))
    // الواضح يتحسب بنوعه المقترح، بنفس قاعدة الرئيسية (OVERRIDES §18)
    const names = new Map(categories.map((c) => [c.id, c.name]))
    const estimated = withEstimatedKinds(transactions, names)
    const counted = estimated.transactions

    const totals = computePeriodTotals(counted, allocations)
    const coverage = assessCoverage(counted)
    const spentMinor = totals.personalExpenseMinor
    // مجهول لا صفر: فيه عمليات وولا واحدة محددة النوع
    const spentKnown = !(coverage.total > 0 && coverage.unclassified === coverage.total)

    const categoryBudgets = budget ? await deps.budgets.listCategoryBudgets(budget.id) : []

    // ─── تاريخ الفترات السابقة: للمتوسط والشذوذ فقط ───
    const history: CompletedPeriodSpend[] = []
    const historyByCategory = new Map<Id, Halalas[]>()

    const historyRows = await Promise.all(Array.from({ length: HISTORY_PERIODS }, async (_, index) => {
      const i = index + 1
      const p = shiftPeriod(period, -i, payday)
      const rows = await deps.txns.listByDateRange(p.start, p.end)
      const rowAllocations = await deps.allocations.listByTransactionIds(rows.map((r) => r.id))
      return { p, rows, rowAllocations }
    }))
    for (const { p, rows: rawRows, rowAllocations } of historyRows) {
      const rows = withEstimatedKinds(rawRows, names).transactions
      const t = computePeriodTotals(rows, rowAllocations)
      const c = assessCoverage(rows)

      history.push({
        periodKey: p.key,
        spentMinor: t.personalExpenseMinor,
        // فترة فيها عملية واحدة غير محددة **ليست موثوقة**
        reliable: c.unclassified === 0,
        transactionCount: rows.length,
      })

      // تاريخ كل تصنيف — من الفترات الموثوقة فقط
      if (c.unclassified === 0 && rows.length > 0) {
        const { slices } = categoryDistribution(rows, rowAllocations)
        for (const slice of slices) {
          if (slice.categoryId === null) continue
          const list = historyByCategory.get(slice.categoryId) ?? []
          list.push(slice.amountMinor)
          historyByCategory.set(slice.categoryId, list)
        }
      }
    }

    const average = averageCompletedSpend(history)

    // متوسط كل تصنيف — من تاريخه هو
    const averageByCategory = new Map<Id, Halalas | null>()
    for (const [categoryId, values] of historyByCategory) {
      averageByCategory.set(
        categoryId,
        averageCompletedSpend(
          values.map((v, i) => ({
            periodKey: `${categoryId}-${i}`,
            spentMinor: v,
            reliable: true,
            transactionCount: 1,
          })),
        ).averageMinor,
      )
    }

    const { slices } = categoryDistribution(counted, allocations)
    const spendByCategory = new Map<Id, Halalas>()
    for (const slice of slices) {
      if (slice.categoryId !== null) spendByCategory.set(slice.categoryId, slice.amountMinor)
    }

    const limitByCategory = new Map(
      categoryBudgets.map((cb) => [
        cb.categoryId,
        { limitMinor: cb.limitMinor, thresholdPercent: cb.notifyEnabled ? cb.thresholdPercent : null },
      ]),
    )

    const totalLimit = budget?.totalLimitMinor ?? null

    return {
      period,
      budget,
      // بلا مصروف معروف لا تُبنى حالة سقف: «0 من 3000» راحة كاذبة
      totalStatus:
        totalLimit === null || !spentKnown
          ? null
          : budgetStatus(totalLimit, spentMinor, budget!.thresholdPercent),
      spentMinor,
      spentReliable: coverage.totalsReliable && estimated.estimatedCount === 0,
      spentNote: estimated.needsReviewCount > 0
        ? `تقريبي: ${estimated.needsReviewCount} عملية محتاجة تأكيد نوعها، والرقم هيتظبط لما تأكدها.`
        : estimated.estimatedCount > 0
          ? `تقريبي: ${estimated.estimatedCount} عملية نوعها اتحدد تلقائي.`
          : coverage.note,
      spentKnown,
      average,
      anomaly: detectAnomaly(
        spentMinor,
        history.filter((h) => h.reliable && h.transactionCount > 0).map((h) => h.spentMinor),
      ),
      // ⚠️ المتاح اليومي يستعمل **السقف** لا المتوسط. لو غاب السقف يبقى غير متاح
      allowance: dailyAllowance(spentKnown ? totalLimit : null, spentMinor, today, period),
      categories,
      lines: buildCategoryLines(spendByCategory, limitByCategory, averageByCategory, historyByCategory),
      categoryBudgets,
    }
  }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/LoadBudgetScreen.kt
package app.masroufy.usecase

import app.masroufy.core.AnomalyResult
import app.masroufy.core.AverageResult
import app.masroufy.core.Budget
import app.masroufy.core.Category
import app.masroufy.core.CategoryBudget
import app.masroufy.core.CategoryBudgetLine
import app.masroufy.core.CategoryLimit
import app.masroufy.core.CompletedPeriodSpend
import app.masroufy.core.DailyAllowance
import app.masroufy.core.Halalas
import app.masroufy.core.Id
import app.masroufy.core.Period
import app.masroufy.core.TextKey
import app.masroufy.core.assessCoverage
import app.masroufy.core.averageCompletedSpend
import app.masroufy.core.budgetStatus
import app.masroufy.core.buildCategoryLines
import app.masroufy.core.categoryDistribution
import app.masroufy.core.computePeriodTotals
import app.masroufy.core.dailyAllowance
import app.masroufy.core.detectAnomaly
import app.masroufy.core.uiText
import app.masroufy.core.withEstimatedKinds
import app.masroufy.port.AllocationRepository
import app.masroufy.port.BudgetRepository
import app.masroufy.port.CategoryRepository
import app.masroufy.port.TransactionRepository

/**
 * شاشة الميزانية — نقل `src/application/useCases/loadBudgetScreen.ts` (spec/01).
 * «سقف إجمالي وسقوف اختيارية للتصنيفات، ومتوسط **منفصل** من فترات مكتملة، وعلامة شذوذ.
 *  لا تُخلق ميزانيات من متوسطات دون اختيار المستخدم.»
 * «منفصل» حرفيًا: `average` و`totalStatus` حقلين مختلفين، والأول عمره ما بيبقى قيمة افتراضية للتاني.
 */

/** كام فترة سابقة بتتقرا للمتوسط والشذوذ. */
private const val HISTORY_PERIODS = 6

data class BudgetScreenData(
    val period: Period,
    val budget: Budget?,
    /** null = المستخدم ما حطش سقف إجمالي. */
    val totalStatus: app.masroufy.core.BudgetStatus?,
    val spentMinor: Halalas,
    /** موثوقية مصروف الفترة الجارية. */
    val spentReliable: Boolean,
    val spentNote: String?,
    /**
     * المصروف **معروف أصلًا**؟ `false` لما ولا عملية محددة النوع: ساعتها الصفر الحسابي
     * **مجهول** مش صفر، وشريط «0 من 3000» راحة كاذبة (CLAUDE.md #10).
     */
    val spentKnown: Boolean,
    /** **معلومة بتتعرض، مش سقف بيتطبق.** */
    val average: AverageResult,
    val anomaly: AnomalyResult,
    val allowance: DailyAllowance,
    val categories: List<Category>,
    val lines: List<CategoryBudgetLine>,
    val categoryBudgets: List<CategoryBudget>,
)

data class LoadBudgetScreenDeps(
    val txns: TransactionRepository,
    val categories: CategoryRepository,
    val allocations: AllocationRepository,
    val budgets: BudgetRepository,
)

data class LoadBudgetScreenRequest(
    val period: Period,
    val today: String,
    val payday: Int,
)

class LoadBudgetScreen(private val deps: LoadBudgetScreenDeps) {
    suspend fun load(request: LoadBudgetScreenRequest): BudgetScreenData {
        val period = request.period
        val transactions = deps.txns.listByDateRange(period.start, period.end)
        val categories = deps.categories.listAll()
        val budget = deps.budgets.findByPeriod(period.key)
        val allocations = deps.allocations.listByTransactionIds(transactions.map { it.id })

        // الواضح بيتحسب بنوعه المقترح، بنفس قاعدة الرئيسية (OVERRIDES §18)
        val names = categories.associate { it.id to it.name }
        val estimated = withEstimatedKinds(transactions, names)
        val counted = estimated.transactions

        val totals = computePeriodTotals(counted, allocations)
        val coverage = assessCoverage(counted)
        val spentMinor = totals.personalExpenseMinor
        // مجهول مش صفر: فيه عمليات وولا واحدة محددة النوع
        val spentKnown = !(coverage.total > 0 && coverage.unclassified == coverage.total)

        val categoryBudgets = budget?.let { deps.budgets.listCategoryBudgets(it.id) } ?: emptyList()

        // ─── تاريخ الفترات السابقة: للمتوسط والشذوذ بس ───
        val history = mutableListOf<CompletedPeriodSpend>()
        val historyByCategory = LinkedHashMap<Id, MutableList<Halalas>>()

        for (i in 1..HISTORY_PERIODS) {
            val p = shiftPeriod(period, -i, request.payday)
            val rawRows = deps.txns.listByDateRange(p.start, p.end)
            val rowAllocations = deps.allocations.listByTransactionIds(rawRows.map { it.id })
            val rows = withEstimatedKinds(rawRows, names).transactions
            val t = computePeriodTotals(rows, rowAllocations)
            val c = assessCoverage(rows)

            history += CompletedPeriodSpend(
                periodKey = p.key,
                spentMinor = t.personalExpenseMinor,
                // فترة فيها عملية واحدة غير محددة **مش موثوقة**
                reliable = c.unclassified == 0,
                transactionCount = rows.size,
            )

            // تاريخ كل تصنيف — من الفترات الموثوقة بس
            if (c.unclassified == 0 && rows.isNotEmpty()) {
                for (slice in categoryDistribution(rows, rowAllocations).slices) {
                    val categoryId = slice.categoryId ?: continue
                    historyByCategory.getOrPut(categoryId) { mutableListOf() }.add(slice.amountMinor)
                }
            }
        }

        val average = averageCompletedSpend(history)

        // متوسط كل تصنيف — من تاريخه هو
        val averageByCategory = LinkedHashMap<Id, Halalas?>()
        for ((categoryId, values) in historyByCategory) {
            averageByCategory[categoryId] = averageCompletedSpend(
                values.mapIndexed { i, v -> CompletedPeriodSpend("$categoryId-$i", v, reliable = true, transactionCount = 1) },
            ).averageMinor
        }

        val spendByCategory = LinkedHashMap<Id, Halalas>()
        for (slice in categoryDistribution(counted, allocations).slices) {
            val categoryId = slice.categoryId ?: continue
            spendByCategory[categoryId] = slice.amountMinor
        }

        val limitByCategory = LinkedHashMap<Id, CategoryLimit>()
        for (cb in categoryBudgets) {
            limitByCategory[cb.categoryId] = CategoryLimit(cb.limitMinor, if (cb.notifyEnabled) cb.thresholdPercent else null)
        }

        val totalLimit = budget?.totalLimitMinor

        return BudgetScreenData(
            period = period,
            budget = budget,
            // من غير مصروف معروف ما تتبنيش حالة سقف: «0 من 3000» راحة كاذبة
            totalStatus = if (totalLimit == null || !spentKnown) null else budgetStatus(totalLimit, spentMinor, budget!!.thresholdPercent),
            spentMinor = spentMinor,
            spentReliable = coverage.totalsReliable && estimated.estimatedCount == 0,
            spentNote = when {
                estimated.needsReviewCount > 0 -> uiText(TextKey.BUDGET_SPENT_NEEDS_REVIEW, estimated.needsReviewCount.toString())
                estimated.estimatedCount > 0 -> uiText(TextKey.BUDGET_SPENT_ESTIMATED, estimated.estimatedCount.toString())
                else -> coverage.note
            },
            spentKnown = spentKnown,
            average = average,
            anomaly = detectAnomaly(
                spentMinor,
                history.filter { it.reliable && it.transactionCount > 0 }.map { it.spentMinor },
            ),
            // ⚠️ المتاح اليومي بيستعمل **السقف** مش المتوسط. لو السقف غايب يبقى غير متاح
            allowance = dailyAllowance(if (spentKnown) totalLimit else null, spentMinor, request.today, period),
            categories = categories,
            lines = buildCategoryLines(spendByCategory, limitByCategory, averageByCategory, historyByCategory),
            categoryBudgets = categoryBudgets,
        )
    }
}
